File: src/ruby/tormenta20/models/regra.rb
# frozen_string_literal: true

require "active_record"
require "json"

module Tormenta20
  module Models
    # Model representing a Regra (Rule) in Tormenta20.
    #
    # Rules contain structured data about game mechanics such as
    # skills, conditions, combat rules, and other game systems.
    #
    # @example Find all rules
    #   Tormenta20.regras.all
    #
    # @example Find a specific rule
    #   Tormenta20.regras.find("pericias")
    #
    # @!attribute [rw] id
    #   @return [String] Unique identifier (e.g., "pericias")
    # @!attribute [rw] name
    #   @return [String] Rule name in Portuguese
    # @!attribute [rw] description
    #   @return [String, nil] Rule description
    # @!attribute [rw] data
    #   @return [Hash, Array] Structured rule data (JSON)
    class Regra < Base
      self.table_name = "regras"

      validates :id, presence: true, uniqueness: true
      validates :name, presence: true
      validates :data, presence: true

      # Convert the rule to a hash representation.
      #
      # @return [Hash] Hash with all rule attributes (nil values excluded)
      def to_h
        {
          id: id,
          name: name,
          description: description,
          data: data
        }.compact
      end
    end
  end
end

File: src/ts/models/regra.ts
import { BaseModel } from "./base.js"
import { Query } from "../query.js"

const TABLE = "regras"

/**
 * A Regra (game rule entry) in Tormenta20.
 *
 * Rules entries contain structured reference data for T20 mechanics.
 *
 * @example
 * ```ts
 * Regra.all()
 * Regra.find("combate")
 * ```
 */
export class Regra extends BaseModel {
  get description(): string | null { return (this._row.description as string) ?? null }
  /** JSON object with structured rule data (schema varies per entry). */
  get data(): Record<string, unknown> { return BaseModel.json(this._row.data, {}) }

  toH() {
    return { id: this.id, name: this.name, description: this.description, data: this.data }
  }

  static query(): Query<Regra> { return BaseModel.makeQuery(TABLE, Regra) }
  static all(): Regra[] { return Regra.query().all() }
  static find(id: string): Regra | null { return Regra.query().find(id) }
  static first(): Regra | null { return Regra.query().first() }
  static last(): Regra | null { return Regra.query().last() }
  static count(): number { return Regra.query().count() }
}
